editor/components/preserve-scroll-in-reorder/index.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+import { withSelect } from '@wordpress/data';
+import { getScrollContainer } from '@wordpress/utils';
+
+/**
+ * Internal dependencies
+ */
+import { getBlockDOMNode } from '../../utils/dom';
+
+/**
+ * Non-visual component which preserves offset of selected block within nearest
+ * scrollable container while reordering.
+ *
+ * @example
+ *
+ * ```jsx
+ * <PreserveScrollInReorder />
+ * ```
+ */
+class PreserveScrollInReorder extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.bindNode = this.bindNode.bind( this );
+	}
+
+	bindNode( node ) {
+		this.node = node;
+	}
+
+	componentWillUpdate( nextProps ) {
+		const { blockOrder, selectionStart } = nextProps;
+		if ( blockOrder !== this.props.blockOrder && selectionStart ) {
+			this.setPreviousOffset( selectionStart );
+		}
+	}
+
+	componentDidUpdate() {
+		if ( this.previousOffset ) {
+			this.restorePreviousOffset();
+		}
+	}
+
+	/**
+	 * Given the block UID of the start of the selection, saves the block's
+	 * top offset as an instance property before a reorder is to occur.
+	 *
+	 * @param {string} selectionStart UID of selected block.
+	 */
+	setPreviousOffset( selectionStart ) {
+		const blockNode = getBlockDOMNode( selectionStart );
+		if ( ! blockNode ) {
+			return;
+		}
+
+		this.previousOffset = blockNode.getBoundingClientRect().top;
+	}
+
+	/**
+	 * After a block reordering, restores the previous viewport top offset.
+	 */
+	restorePreviousOffset() {
+		const { selectionStart } = this.props;
+		const scrollContainer = getScrollContainer( this.node );
```

Comment:
The question I'm wondering is whether we should preserve the scrolling relatively to the scrollable parent of `blockNode` or the parent of `this.node`. I'm thinking maybe the most logical would be `blockNode` even if it's the same node right now.

Reply:
>I'm thinking maybe the most logical would be `blockNode` even if it's the same node right now.

Good point. Also helps simplify the component by removing the need for _any_ return value or node binding. Updated in rebased eb63976.